Address performance issues in the AnalyzerManager reported at…

mavasani opened the pull request:
…https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/2078#issuecomment-181150995:

1. Replace use of ConcurrentDictionaries with Dictionary + lock.
2. Avoid unncessary ToImmutableArray invocations.

I also fixed up a subtle race condition in CompilationAnalysisValueProvider.TryGetValue for my last PR (see https://github.com/dotnet/roslyn/pull/8375/files#r52127390).

mavasani commented:
Tagging @JohnHamby @dotnet/roslyn-analysis @dotnet/roslyn-compiler

mavasani commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationAnalysisValueProvider.cs:
```diff
@@ -42,7 +42,18 @@ internal bool TryGetValue(TKey key, out TValue value)
             lock (_valueMap)
             {
-                _valueMap[key] = value;
+                // Check if another thread already stored the computed value.
```
@sharwell This should address your comment: https://github.com/dotnet/roslyn/pull/8375/files#r52127390

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:
```diff
@@ -32,21 +32,44 @@ internal partial class AnalyzerManager
+                    return value;
+                }
+
+                _compilationScopeMap[analyzerAndOptions] = value = new ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>>();
```
Separate assignments to different lines.

cston replied:
Consider using ```_compilationScopeMap.Add``` instead so it's clear the key should not exist in the map. Same comment for `_sessionScopeMap` below.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:
```diff
@@ -168,49 +207,84 @@ internal partial class AnalyzerManager
+                if (_descriptorMap.TryGetValue(analyzer, out storedValue))
+                {
+                    // If so, we return the stored value.
+                    value = storedValue;
```
Are we testing this case?

mavasani replied:
Unfortunately, it won't be easy to test it as it will require 2 analysis threads to simultaneously request descriptors for the same analyzer, such that both of them go past the first lock before any of them enter the second lock.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationAnalysisValueProvider.cs:
```diff
@@ -42,7 +42,18 @@ internal bool TryGetValue(TKey key, out TValue value)
+                if (_valueMap.TryGetValue(key, out storedValue))
+                {
+                    // If so, we return the stored value.
+                    value = storedValue;
```
Consider adding a test.

mavasani replied:
I can add a defensive test where multiple threads try to fetch the value from the same provider for a specific compilation.

mavasani commented:
I have addressed the PR feedback. 

@cston I haven't added unit tests as there are no changes in this PR that are needed for functional correctness, they are purely for performance reasons. Even the change to CompilationAnalysisValueProvider is not required per spec, but is just a better implementation (spec does not guarantee reference equality of values even within a compilation context).

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:
```diff
@@ -168,49 +209,84 @@ internal partial class AnalyzerManager
+                else
+                {
+                    // Otherwise, store the value computed here.
+                    _descriptorMap[analyzer] = value;
```
Consider ```_descriptorMap.Add(analyzer, value)``` instead. Same comment for ```_valueMap.Add(key, value)``` below.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:
```diff
@@ -221,47 +297,80 @@ internal void ClearAnalyzerState(ImmutableArray<DiagnosticAnalyzer> analyzers)
+                            {
+                                descriptor.Title.OnException -= handler;
+                                descriptor.MessageFormat.OnException -= handler;
+                                descriptor.Description.OnException -= handler;
```
Do we have implementations of `LocalizableString` that throw exceptions? Is this a class can be implemented by 3rd parties?

mavasani replied:
Yes, arguments to the analyzer's supported descriptors are of `LocalizableString` type, which is a public abstract type. We do provide an exception safe [default implementation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/LocalizableResourceString.cs,5febf21f2f3ec86c,references) based on resource strings, but 3rd parties can provide their own implementation which can throw. We need to catch those exceptions and log them as analyzer exception diagnostics.

Unit tests for verifying this scenario: http://source.roslyn.io/#Roslyn.Compilers.UnitTests/Diagnostics/DiagnosticLocalizationTests.cs,12bc2060edf8e15d

cston commented:
LGTM

mavasani commented:
Failure is due to known issue: https://github.com/dotnet/roslyn/issues/8077

mavasani commented:
@dotnet-bot retest prtest/win/dbg/unit64 please

mavasani commented:
Seems that prtest/determinism is stuck - ignoring that as per Jared's email.
